cunumeric/thunk.py

```diff
@@ -31,17 +34,19 @@ class NumPyThunk(ABC):
     :meta private:
     """
 
-    def __init__(self, runtime, dtype) -> None:
+    array: Any
```

Comment:
`DeferredArray` objects don't have an `array` attribute, only `EagerArray` objects.

Reply:
Removing this yields
```
~/work/cunumeric bryanv/mypy_thunks*
gas38 ❯ mypy ../legate.core/install38/lib/python3.8/site-packages/legate /home/bryan/work/legate.core/typings/ cunumeric
cunumeric/eager.py: note: In member "bincount" of class "EagerArray":
cunumeric/eager.py:627:17: error: "NumPyThunk" has no attribute "array"  [attr-defined]
                    weights.array if weights is not None else None,
                    ^
Found 1 error in 1 file (checked 83 source files)
```
Where:
```
    def bincount(self, rhs: Any, weights: Optional[NumPyThunk] = None) -> None:
        self.check_eager_args(rhs, weights)
        if self.deferred is not None:
            self.deferred.bincount(rhs, weights=weights)
        else:
            self.array[:] = np.bincount(
                rhs.array,
                weights.array if weights is not None else None,
                minlength=self.array.size,
            )
```
Does eager `bincount` only accept deferred `weights`? Is this a bug?

Reply:
So actually all `EagerArray` thunk methods should accept both `DeferredArray` and `EagerArray` objects. However, the `check_eager_args` call at the beginning of every such method will fill `self.deferred` if any of the arguments are `DeferredArray`s, thus guaranteeing that the "deferred fallback" path will be taken (`if self.deferred is not None`). That path will just pass the arguments to the corresponding `DeferredArray` thunk method, and all such methods are supposed to also accept both `DeferredArray` and `EagerArray` objects (because they get converted by the `@auto_convert` decorator).

The invariant here is that if an `EagerArray` object has `self.deferred` filled, then every operation should be forwarded to that.

If you can think of a saner way to encode the same behavior, we could tackle it in a subsequent PR (and just leave the fake `array` member, or any other similar workaround, in for now, with a comment that it's there to work around convoluted typing behavior).

P.S. The above discussion also explains why the `NumPyThunk` methods should be able to accept any argument of `NumPyThunk` type (or, in some cases, `NumPyThunk | None`).

P.P.S. As I was looking into this, I found errors in the use of `check_eager_args` in places. I will push fixes on a small PR.

Reply:
@manopapad I might be misunderstanding, but the problem seems to be in the other branch, when `self.deferred` *is* None. In that case the codepath expects `weights.array` to exist. Are you saying that branch should just never been taken?

Reply:
Actually I just realized I read things backwards, i.e. that deferred array's have the `array` I think the simplest "fix" here for now is just a cast

Reply:
The end result of all the above is that the `then` branch is taken only if all thunk-type arguments are `EagerArray` or `None`.

Reply:
Yup, I read your initial comment exactly backwards :astonished:

Reply:
https://github.com/nv-legate/cunumeric/pull/438/commits/51424b0066e5922ab6d9c411abce6b745b6f2a92